src/provider/store/Store.ts

```diff
@@ -1,122 +1,89 @@
+}
+
+type Listener<S> = (getState: () => S) => void;
+
+export class Store<S, A extends Action<S>> extends AsyncInit implements StoreAPI<S, A> {
+    private _currentState: S;
+    private _listeners: Listener<S>[] = [];
 
-    protected async init(): Promise<void> {
-        await this._database.initialized;
-        this._store = createStore<RootState, RootAction, {}, {}>(this.reduce.bind(this), await this.getInitialState(), this.createEnhancer());
+    public readonly onAction: Signal<[A], Promise<void>, Promise<void>> = new Signal(Aggregators.AWAIT_VOID);
+
+    constructor(initialState: S) {
+        super();
+        this._currentState = initialState;
     }
 
-    public get state(): RootState {
-        return this._store.getState() as RootState;
+    public get state(): S {
+        return this._currentState;
     }
 
-    public async dispatch(action: RootAction): Promise<void> {
-        if (action instanceof CustomAction) {
+    public async dispatch(action: A): Promise<void> {
+        if (action instanceof AsyncAction) {
             // Action has custom dispatch logic
             await action.dispatch(this);
-        } else {
-            // Pass straight through to redux store
-            await this._store.dispatch({...action});
         }
+        return this.reduceAndSignal(action);
     }
 
-    public watchForChange<T>(getObject: (state: RootState) => T, observer: StoreChangeObserver<T>): Unsubscribe {
-        const watcher = this.watch<T>(() => this.state, getObject);
-        return this._store.subscribe(watcher(observer));
-    }
+    public subscribe(listener: Listener<S>) {
+        this._listeners.push(listener);
 
-    private watch<T>(getState: () => any, getObject: (state: RootState) => T): (observer: StoreChangeObserver<T>) => () => void {
-        let currentValue = getObject(getState());
-        return function w(observer: StoreChangeObserver<T>) {
-            return function () {
-                const newValue: T = getObject(getState());
-                if (currentValue !== newValue) {
-                    const oldValue = currentValue;
-                    currentValue = newValue;
-                    observer(oldValue, newValue);
-                }
-            };
+        return () => {
+            const index: number = this._listeners.indexOf(listener);
+            if (index >= 0) {
+                this._listeners.splice(index, 1);
+            }
         };
     }
 
-    private reduce<T extends Action>(state: RootState | undefined, action: ActionOf<T>): RootState {
-        const handler: ActionHandler<T> | undefined = this._actionHandlerMap[action.type] as ActionHandler<T>;
+    protected async init(): Promise<void> {}
 
-        if (handler) {
-            return handler(state!, action);
-        } else {
-            // No handler registered for this action - action does not modify the store's state
-            return state!;
-        }
+    protected setState(state: S): void {
```

Comment:
Any thoughts on making `protected setState(state: S)` into `protected set state(state: S)`.  So in our code, we always use `.state` for both setting and getting?

Reply:
then the getter and setter won't agree in visibility. we don't wanna expose the setter

Reply:
👍